orocos_kdl/src/chainexternalwrenchestimator.cpp

```diff
@@ -0,0 +1,196 @@
+// Calculates robot's initial momentum in the joint space. If this method is not called by the user, zero values will be taken for the initial momentum.
+int ChainExternalWrenchEstimator::setInitialMomentum(const JntArray &joint_position, const JntArray &joint_velocity)
+{
+    // Check sizes first
+    if (joint_position.rows() != nj || joint_velocity.rows() != nj) return (error = E_SIZE_MISMATCH);
+
+    // Calculate robot's inertia and momentum in the joint space
+    dynparam_solver.JntToMass(joint_position, jnt_mass_matrix);
+    initial_jnt_momentum.data = jnt_mass_matrix.data * joint_velocity.data;
+
+    // Reset data because of the new momentum offset
+    SetToZero(estimated_momentum_integral);
+    SetToZero(filtered_estimated_ext_torque);
+
+    return (error = E_NOERROR);
+}
+
+// Sets singular-value eps parameter for the SVD calculation
+void ChainExternalWrenchEstimator::setSVDEps(const double eps_in)
+{
+    svd_eps = eps_in;
+}
+
+// Sets maximum iteration parameter for the SVD calculation
+void ChainExternalWrenchEstimator::setSVDMaxIter(const int maxiter_in)
+{
+    svd_maxiter = maxiter_in;
+}
+
+// This method calculates the external wrench that is applied on the robot's end-effector.
+int ChainExternalWrenchEstimator::JntToExtWrench(const JntArray &joint_position, const JntArray &joint_velocity, const JntArray &joint_torque, Wrench &external_wrench)
+{
+    /**
+     * ==========================================================================
+     * First-order momentum observer, an implementation based on:
+     * S. Haddadin, A. De Luca and A. Albu-Schäffer,
+     * "Robot Collisions: A Survey on Detection, Isolation, and Identification,"
+     * in IEEE Transactions on Robotics, vol. 33(6), pp. 1292-1312, 2017.
+     * ==========================================================================
+     */
+
+    // Check sizes first
+    if (nj != CHAIN.getNrOfJoints() || ns != CHAIN.getNrOfSegments()) return (error = E_NOT_UP_TO_DATE);
+    if (joint_position.rows() != nj || joint_velocity.rows() != nj || joint_torque.rows() != nj) return (error = E_SIZE_MISMATCH);
+
+    /**
+     * =======================================================================================================================
+     * Part I: Estimation of the torques that are felt in joints as a result of the external wrench being applied on the robot
+     * =======================================================================================================================
+     */
+
+    // Calculate decomposed robot's dynamics
+    int solver_result = dynparam_solver.JntToMass(joint_position, jnt_mass_matrix);
+    if (solver_result != 0) return solver_result;
+    solver_result = dynparam_solver.JntToCoriolis(joint_position, joint_velocity, coriolis_torque);
+    if (solver_result != 0) return solver_result;
+    solver_result = dynparam_solver.JntToGravity(joint_position, gravity_torque);
+    if (solver_result != 0) return solver_result;
```

Comment:
Yes, they can. Here, there are three possible returns:
* //! Chain size changed
        E_NOT_UP_TO_DATE = -3,
* //! Input size does not match internal state
        E_SIZE_MISMATCH = -4,
* //! No error
        E_NOERROR          =  0.


True, I agree about debugging. What is your suggestion for addressing this problem? SolverI class only contains generic error definitions, but not specific for this solver... Should I introduce some additional error definitions in SolverI or in the Estimator class?

Reply:
I think a print statement would already be a good improvement.

Reply:
I would avoid adding print statements directly in the solvers as much as possible.  Even in case of error, this will break realtime usage.  @DjoleMNE afaik there are already solvers that add additional solver specific error definitions, e.g. https://github.com/orocos/orocos_kinematics_dynamics/blob/master/orocos_kdl/src/chainiksolverpos_nr.hpp#L41-L42, https://github.com/orocos/orocos_kinematics_dynamics/blob/master/orocos_kdl/src/chainiksolverpos_nr.cpp#L74-L79

Reply:
Great, thanks for the feedback. I will work on it.

Reply:
Done.